### PR TITLE
Remote publication using min node version for backward compatibility

### PR DESCRIPTION
**Pull request by soosinha**
### Description
- Remote cluster state state should be published using min node version in the cluster. If even a single node is not remote enabled then remote publication should not be used.

### Related Issues
NA

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java [hunk 1]
```diff
@@ -443,7 +444,8 @@ public RemoteClusterStateManifestInfo writeIncrementalMetadata(
             uploadedMetadataResults,
             previousManifest.getPreviousClusterUUID(),
             clusterStateDiffManifest,
-            false
+            false,
+            previousManifest.getCodecVersion()
```
Comment (himshikha) [inline server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:448]: What happens if between the previous manifest and current upload, a node with lower version gets added to the cluster?

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java [hunk 2]
```diff
@@ -332,9 +332,18 @@ public PublicationContext newPublicationContext(
...
+        if (isRemotePublicationEnabled == true) {
+            for (DiscoveryNode node : clusterChangedEvent.state().nodes().getNodes().values()) {
+                // if a node is non-remote then created local publication context
+                if (node.isRemoteStatePublicationEnabled() == false) {
+                    break;
+                }
+                // if all nodes are remote then create remote publication context
+                return new RemotePublicationContext(clusterChangedEvent, persistedStateRegistry);
+            }
+        }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:344]: WOuldn't this for loop do only the first iteration and either `break` or  `return`

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java (hunk 1 above)
Comment (soosinha) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:448]: if a node is added to the cluster, cluster state will get updated and manifest will also change accordingly

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java (hunk 2 above)
Comment (soosinha) [reply server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:344]: Yes. Will move this out of the loop

File: server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java (hunk 1 above)
Comment (himshikha) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:448]: But we will still use codec from previous manifest, can that lead to incorrect version being picked?

Comment (soosinha) [reply server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java:448]: If a node with lower version gets added and the codec for that version is also lower than current manifest, then will write full cluster state. Incremental cluster state will be written when codec version in previous manifest and that for current cluster state nodes is the same.

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java [hunk 3]
```diff
@@ -332,11 +332,13 @@ public PublicationContext newPublicationContext(
         boolean isRemotePublicationEnabled,
         PersistedStateRegistry persistedStateRegistry
     ) {
-        final PublicationContext publicationContext = new PublicationContext(
-            clusterChangedEvent,
-            isRemotePublicationEnabled,
-            persistedStateRegistry
-        );
+        if (isRemotePublicationEnabled == true) {
+            if (validateRemotePublicationOnAllNodes(clusterChangedEvent.state().nodes()) == true) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:336]: this check will be triggered with every state update and would be un-necessary unless nodes are joining the cluster. Also once cluster moves to remote publication if one node joins with state publication disabled would it be allowed? It shouldn't be. Once a cluster starts doing remote publishing, it shouldn't go back and forth with local/remote publication dynamically.

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java [hunk 4]
```diff
@@ -648,4 +599,63 @@ public String executor() {
...
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        transportExceptionHandler.accept(exp);
+                    }
+
+                    @Override
+                    public String executor() {
+                        return ThreadPool.Names.GENERIC;
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:644]: should we move state publishing to different threadpool?

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java [hunk 5]
```diff
@@ -758,11 +762,11 @@ private boolean verifyManifestAndClusterState(ClusterMetadataManifest manifest,
             return true;
         }
 
-        private boolean shouldWriteFullClusterState(ClusterState clusterState) {
+        private boolean shouldWriteFullClusterState(ClusterState clusterState, int codecVersion) {
             if (lastAcceptedState == null
                 || lastAcceptedManifest == null
                 || lastAcceptedState.term() != clusterState.term()
-                || lastAcceptedManifest.getOpensearchVersion() != Version.CURRENT) {
+                || lastAcceptedManifest.getCodecVersion() != codecVersion) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:769]: lets add assert that lastAcceptedManifest.getCodecVersion() <= codecVersion

File: server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java [hunk 6]
```diff
@@ -245,6 +258,20 @@ private static ClusterStateDiffManifest diffManifest(Object[] fields) {
         declareParser(PARSER_V3, CODEC_V3);
     }
 
+    public static int getCodecForVersion(Version version) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java:261]: would it be simple to generate a map from the CODEC_TO_VERSION_MAPPING for every version up to the current version. This way we dont have to run a for loop for every manifest.

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java (hunk 3 above)
Comment (soosinha) [reply server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:336]: What if the user wants to disable the remote publication. New nodes will join with remote publication disabled. At this point we should stop remote publishing, right ?

Comment (shwetathareja) [reply server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:336]: ideally remote publishing should be controlled by cluster manager instead of individual nodes. CM decides when to enable/ disable using the setting. This min compatibility check is to ensure along with setting, during upgrades, it can smoothly enable it without triggering downtime but it is not an expectation during regular operation of a running cluster.

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java (hunk 4 above)
Comment (shwetathareja) [reply server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:644]: Lets create a tracking issue.

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java (hunk 5 above)
Comment (shwetathareja) [reply server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:769]: once we move to higher codec version, it can't move back to lower codec version automatically. Either this should be a static setting passed explicitly to cluster manager. There is a risk of creating these as dynamic setting as it can mess up with cluster state.

File: server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java (hunk 6 above)
Comment (shwetathareja) [reply server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java:261]: you can auto generate the map, it need not be hard coded.

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java (hunk 5 above)
Comment (soosinha) [reply server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:771]: Added assert

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java [hunk 7]
```diff
@@ -345,6 +354,19 @@ public PublicationContext newPublicationContext(
         return publicationContext;
     }
 
+    private boolean validateRemotePublicationOnAllNodes(DiscoveryNodes discoveryNodes) {
+        if (ClusterMetadataManifest.getCodecForVersion(discoveryNodes.getMinNodeVersion()) < ClusterMetadataManifest.CODEC_V0) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:358]: when will this be true?

Comment (soosinha) [reply server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:358]: This shouldn't be true. I can add an assert instead

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java [hunk 8]
```diff
@@ -758,11 +763,12 @@ private boolean verifyManifestAndClusterState(ClusterMetadataManifest manifest,
             return true;
         }
 
-        private boolean shouldWriteFullClusterState(ClusterState clusterState) {
+        private boolean shouldWriteFullClusterState(ClusterState clusterState, int codecVersion) {
+            assert lastAcceptedManifest == null || lastAcceptedManifest.getCodecVersion() <= codecVersion;
             if (lastAcceptedState == null
                 || lastAcceptedManifest == null
                 || lastAcceptedState.term() != clusterState.term()
-                || lastAcceptedManifest.getOpensearchVersion() != Version.CURRENT) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:765]: @soosinha I am wondering should we still keep OpenSearchVersion check along with codec as well

**Comment by shwetathareja** [comment]
please add changelog as well

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java (hunk 8 above)
Comment (soosinha) [reply server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:765]: I will add the version check back as this might provide additional safeguard by handling cases where here is change in the cluster state objects in newer version but the codec version is not updated.

File: server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java [hunk 9]
```diff
@@ -237,12 +236,33 @@ private static ClusterStateDiffManifest diffManifest(Object[] fields) {
...
+        Map<Version, Integer> versionToCodecMapping = new HashMap<>();
+        for (Version version : Version.getDeclaredVersions(Version.class)) {
+            if (version.onOrAfter(Version.V_2_10_0) && version.before(Version.V_2_12_0)) {
+                versionToCodecMapping.put(version, ClusterMetadataManifest.CODEC_V0);
+            } else if (version.onOrAfter(Version.V_2_12_0) && version.before(Version.V_2_15_0)) {
+                versionToCodecMapping.put(version, ClusterMetadataManifest.CODEC_V1);
+            } else if (version.onOrAfter(Version.V_2_15_0) && version.before(Version.V_2_16_0)) {
+                versionToCodecMapping.put(version, ClusterMetadataManifest.CODEC_V2);
+            } else if (version.onOrAfter(Version.V_2_16_0)) {
+                versionToCodecMapping.put(version, ClusterMetadataManifest.CODEC_V3);
```
[the remark this answers is not in the document]
Comment (shwetathareja) [reply server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java:258]: You can do this incrementally in next PR as well

[the remark this answers is not in the document]
Comment (soosinha) [reply server/src/main/java/org/opensearch/gateway/remote/ClusterMetadataManifest.java:262]: Added assert

**Review by shwetathareja — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java [hunk 10]
```diff
@@ -332,11 +334,18 @@ public PublicationContext newPublicationContext(
...
+                if (validateRemotePublicationOnAllNodes(clusterChangedEvent.state().nodes()) == true) {
+                    allNodesRemotePublicationEnabled.set(true);
+                }
+            }
+            if (allNodesRemotePublicationEnabled.get() == true) {
+                // if all nodes are remote then create remote publication context
+                return new RemotePublicationContext(clusterChangedEvent, persistedStateRegistry);
+            }
+        }
+        final PublicationContext publicationContext = new PublicationContext(clusterChangedEvent, persistedStateRegistry);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/coordination/PublicationTransportHandler.java:348]: Sorry why is this needed. Can we publish cluster state dual mode based on the destination node attribute?
